conans/client/generators/__init__.py

```diff
@@ -93,7 +93,12 @@ def write_generators(conanfile, hook_manager):
 
     if conanfile.generators:
         conanfile.output.info(f"Writing generators to {new_gen_folder}")
-    for generator_name in set(conanfile.generators):
+    # generators check that they are not present in the generators field,
+    # to avoid duplicates between the generators attribute and the generate() method
+    # They would raise an exception here if we don't invalidate the field while we call them
+    old_generators = set(conanfile.generators)
+    conanfile.generators = []
```

Comment:
maybe it needs a sort of context manager here?
in case exception happens on line 114, `conanfile.generators` will be never restored to an original state and it will remain empty forever

Reply:
A ``try-finally`` would be good, yes.